Differentiate between null data and zero data on the data overlay

nix-chu opened the pull request:
## Description

This PR definitely should not have taken a month to complete, but we here now so yeet. 🎉 

Having worked on this branch on and off during the last month, I am definitely sure that I missed something or there is a more cleaner way to write the code. So please do not hesitate to scrutinise the **hell** out of this PR. 

Merging in this PR should allow the Raspicam to recognise when data has stopped receiving after 5 seconds. When the Raspicam realises that data has not updated after 5 seconds, then the data overlay should display as `--` representing empty data. The reason behind this PR is for the rider to recognise 'no change' in sensor data on the overlay.

For example, if the rider is riding at a constant 50km/h for more than 5 seconds, then it should keep displaying 50km/h. But if the cadence was at 180 revolutions per minutes at one point, and the rider stops pedalling for more than 5 seconds, then it should update that there is no data here because the rider is not pedalling, whereas before it would keep showing 180 revolutions per minute.

## Screenshots

![image](https://user-images.githubusercontent.com/58575113/92999155-3996b680-f562-11ea-8e9d-a78f4dc14b96.png)
Peep my heavy eyebags due to very late nights.... Midsem break cannot come any quicker

## Steps to Test

1. Prepare three terminals, one in the Raspicam directory and the second in the data-acquisition-system/DAS/. Ensure there is a poetry virtual environment in both terminals. For the third terminal, it should run `mosquitto`.
2. In the DAS terminal, run `python das/tests/fake_module_tester.py` to load some fake data onto the mosquitto broker.
3. In the Raspicam terminal, run `python overlay_new.py` (or any of the overlays, but this is the most cleanest looking one imo)
4. Upon opening the Raspicam, there should be data updating onto the overlay. 
5. Stop the `fake_module_tester.py` in the DAS terminal, then quickly switch back onto the Raspicam overlay.
6. After some time (should be 5 seconds), the data points should disappear and appear "--" on the overlay.

nix-chu commented:
Could someone help me figure out why Linting didn't work? 

Also yes, I will download black and flake8. Just need to fix my pyenv first 😅 , tackle one problem at a time is what I've learned when dealing with software systems.

hallgchris commented:
> Could someone help me figure out why Linting didn't work?

Until you've got your environment properly set up (happy to help with this in the drop-in!), you should be able to install `black` globally with `pip install black`, and then run it with `black [python file to format]`. According to the CICD messages, just need to run it on `data_field.py` and `overlay.py` to pass!

hallgchris commented on data.py:
```diff
@@ -1,53 +1,95 @@
+            "rec_speed": DataValue(float),
+            "predicted_max_speed": DataValue(float),
+            "zdist": DataValue(float),
+            "plan_name": DataValue(str),
```
This looks really great! Here's an extension you could do to remove some code you've actually made rather redundant. Right now, messages from the dashboard are stored separately in `self.message` and we store their received time in `self.message_received_time`. Sound familiar? It's basically what you've implemented with `DataField`. So, really we could replace all this logic by storing messages in a data field. Here's what you could do:

1. Make `DATA_EXPIRY` a member variable of `DataValue`, so that it can be set per instance
2. Make data expiry an optional parameter of `DataValue`'s constructor, defaulting to 5 but able to be set to a custom value
3. Replace all the `self.message*` stuff with a single stored instance of a `DataValue`, representing message. Give it a longer expiry timeout, say 20 seconds.
4. Update all the message related functions to just call `DataValue` methods:
    - `load_message` calls `update`
    - `has_message` calls `is_valid`
    - `get_message` calls `get_string`

If you reckon this would take too long or is beyond the scope of this PR, I'm happy for you to leave it for later and just get this PR in. If so, make a notion card and copy this info or a link to the comment in!

nix-chu replied:
Did this but I was thinking if I should just replace all the `load_message` with `update` instead, and also for the others. But since I've never used DAShboard before and don't know how to send messages onto the overlay yet, there's no way for me to test this. So I've just did the method calls

hallgchris replied:
Working brilliantly! You can send messages using only MQTT with
```bash
mosquitto_pub -t "/v3/camera/message" -m '{"message":"hello there"}'
```
although I do remember having some funky issues with quotes on windows, I don't remember which you have to use but I might leave you to figure out what needs doing.

As for running the dashboard, it's fairly straightforward - documented in the readme, but the short of it is
- Install [yarn](https://yarnpkg.com/)
- In one terminal, `cd server`, `yarn install`, `yarn start`
- In another, `cd client`, `yarn install`, `yarn start`

And you should be good to go!

hallgchris replied:
I'm happy with `load_message` still being called that. It means that things interacting with `Data`'s message stuff don't need to worry about the underlying `DataValue` implementation. That's what made this change so easy, you were able to just change how the functions worked internally, and not worry about the stuff that calls the functions at all - things just worked, without even testing!

hallgchris commented on components/centre_power.py:
```diff
@@ -30,13 +30,19 @@ def draw_base(self, canvas: Canvas):
 
-        power_str = f"{power:.0f}w"
-        rec_power_str = f"{rec_power:.0f}w rec"
+        if power is not None:
+            power_str = power + "w"
+        else:
+            power_str = "--"
```
Since we can't do that, this is I reckon the next best thing.
```python
if data["power"].is_valid():
    power_str = data["power"].get_string() + "w"
else
    power_str = "--"
```
By doing this, we never create the potentially invalid variable `power`, which could be misused later if someone didn't know that it *could* be `None`

khanguslee replied:
Excuse me. Please use this by it's more formal terminology. It's called the [walrus operator](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions). Thank you for listening to my TED talk.

hallgchris replied:
I can't believe I forgot about this!

hallgchris commented on pyproject.toml:
```diff
@@ -26,3 +26,20 @@ line-length = 79
 [build-system]
 requires = ["poetry>=0.12"]
 build-backend = "poetry.masonry.api"
+
+[pycalver]
+current_version = "v202009.0001-alpha"
+version_pattern = "{pycalver}"
+commit = true
+tag = true
+push = true
+
+[pycalver.file_patterns]
+"pyproject.toml" = [
+    'current_version = "{version}"',
+]
+"README.md" = [
+    "{version}",
+    "{pep440_version}",
+]
+
```
We're not yet using `pycalver` on this repo. It's not a bad idea for the future, but save for another PR :)

nix-chu replied:
Ah, I did that without thinking cuz GitHub sent the reminder

hallgchris replied:
Hmm, you probably got sent a reminder for a PR in another repo :joy:

hallgchris commented on tests/test_data.py:
```diff
@@ -172,8 +172,8 @@ def test_v3_power_model_parsing():
+        assert data["rec_speed"].get() == 20
+        assert data["zdist"].get() == 30
+        assert data["predicted_max_speed"].get() == 100
+        assert data["plan_name"].get() == "default.pkl"
```
Can you add some `DataValue` tests that test things like
- Initially, `is_valid` is `False` and `get*` return `None`
- After calling `update`, the above functions return correctly
- After waiting 5 seconds, the above functions (or ideally less, just so the tests run faster, if you add something to set a custom data expiry duration) return `False`/`None` again

and in a seperate test,
- Check that the formatting in `get_string` works as expected

hallgchris reviewed (changes requested):
It seems like this PR doesn't run as is, not due to problems introduced here but in a prior PR. For other reviewers, you'll need to update `common` to `202008.6` in `pyproject.toml` and add `()` to the end of line 235 of `backend.py`. These issues are fixed in #45.

Working as expected here - here it is receiving V3 data, but without BOOST running. (Note that elapsed distance is not showing - this is a bug unrelated to the PR, I'll work on a quick fix for that)
![image](https://user-images.githubusercontent.com/17876556/92927910-b140e480-f481-11ea-809d-68ef7d6ff165.png)

Great PR! This is something that raspicam has reeeeaally needed for ages... And you've implemented it really nicely :)

hallgchris commented on data.py:
```diff
@@ -201,24 +243,24 @@ def load_sensor_data(self, data: str) -> None:
-                self.data["reed_velocity"] = float(sensor_value)
-            elif sensor_name in self.data_types:
-                cast_func = self.data_types[sensor_name]
-                self.data[sensor_name] = cast_func(sensor_value)
+                self.data["reed_velocity"].update(float(sensor_value))
```
One quick addition: Can you duplicate these two lines to also update `reed_distance` when we get data on `reedDistance`? Was going to make my own PR but I realise that it would cause a conflict with yours so probably just faster that you add it in here.

nix-chu replied:
Actually, these could be done inside the `else` statement right? Since this is now just the same thing

hallgchris replied:
Not quite the same thing! (annoyingly)

In `self.data`, the keys are names of the V2 data fields which use snake_case, but in V3 we're using CamelCase. (I'm not entirely sure why that was changed... It was though). So, for these multi-word data names, we need to manually map the snake case into camel case.

nix-chu replied:
Ohhh I see the issue now. Haha that's so strange

khanguslee reviewed (commented):
Got no additional comments to make currently! @hallgchris has pretty much pointed out everything!

khanguslee commented:
Might look into doing pre-commit hooks to run `black` so that you don't get the code quality failing. hmm.

nix-chu commented:
Key changes:
- Removed unnecessary indentation in docstrings
- Moved casting logic inside the `DataValue.update()` method
- Wrote test cases for DataValue class
- Got Pyenv For Windows to finally work lmao (there was a short hiccup where changing between Python versions when using the `fake_module_tester.py`, but got that sorted out 👍. Still might be a problem in the future hah lol xd)

hallgchris commented on components/centre_power.py:
```diff
@@ -30,13 +30,19 @@ def draw_base(self, canvas: Canvas):
     def draw_data(self, canvas: Canvas, data: Data):
-        power = data["power"]
-        rec_power = data["rec_power"]
+        power = data["power"].get()
+        rec_power = data["rec_power"].get()
```
Whoops, I thought we'd be able to get rid of these lines, didn't realise we still need the numerical values below. Ah well

hallgchris commented on components/centre_power.py:
```diff
@@ -30,13 +30,19 @@ def draw_base(self, canvas: Canvas):
+            rec_power_str = "--"
 
-        if rec_power == 0:
+        if rec_power == 0 or rec_power is None:
```
Sorry, should have picked this up before. Can you add `or power is None` to this? In the case that `power` is invalid but `rec_power` still has data, we'll divide `None` by a number below which python won't like.

hallgchris commented on data.py:
```diff
@@ -1,92 +1,116 @@
-            return False
-        return True
+        """Assess whether message has expired. Will return true if expiration
+        has not bee exceeded."""
```
Typo: `been`

Out of interest, why were these docstrings changed (`get_message` too)? I'd argue the old ones were slightly more correct/useful. This docstring doesn't consider when no messages have been sent (in which case, they technically haven't expired) and the below docstring talks about `self.message`, which is an internal variable that users of the class shouldn't have to worry about.

hallgchris commented on data.py:
```diff
@@ -201,24 +224,21 @@ def load_sensor_data(self, data: str) -> None:
             if sensor_name == "gps":
-                self.data["gps"] = 1
-                self.data["gps_speed"] = float(sensor_value["speed"])
-            elif sensor_name == "reedVelocity":
-                self.data["reed_velocity"] = float(sensor_value)
```
We still need this! (see discussion in prior review)

hallgchris commented on tests/test_data.py:
```diff
@@ -113,7 +152,7 @@ def test_v3_sensor_parsing():
             "sensors": [
                 {"type": "co2", "value": 35},
-                {"type": "reedVelocity", "value": 20.1},
+                {"type": "reed_velocity", "value": 20.1},
```
This was correct before. V3 mqtt data uses camel case, but in raspicam for legacy reasons we still store the data in snake case. This is imitating MQTT data, so it should be camel case.

hallgchris replied:
Also, to test the reed distance field you're adding, can you add some data to the test case?
```json
{"type": "reedDistance", "value": 1040.3},
```

hallgchris reviewed (changes requested):
Very close! Just minor things to go. Can you also merge in the latest master?

nix-chu commented:
@hallgchris I genuinely believe that Poetry hates me

nix-chu commented:
Ok I have no clue, someone help please lmao

hallgchris reviewed (approved):
Looks great, thanks heaps for this!

nix-chu commented:
@hallgchris whlie this PR is still open for any people wanting to check it out, I've found a way to search for all instances of `Data.load_message()` in a repository and replace them with `DataValue.update()`. Do you think this change should still be required?

hallgchris commented:
Had a think about it, if you want you could change `load_message()` to call `.update()` directly! Not too fussed. However (not sure if you were considering this) I wouldn't change `has_message` and `get_message`, as this would require things outside the class to directly access class members, which is generally best to be avoided. `load_message` is fine because it is only called from within the class or from subclasses anyway.

(e.g. `self.data.get_message()` vs `self.data.message.get()`, best to avoid accessing `message` like that)

nix-chu commented:
Alright that's fair. I won't change it. 👍
